.cbmc-batch/jobs/aws_priority_queue_pop/aws_priority_queue_pop_harness.c

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/priority_queue.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+void aws_priority_queue_pop_harness() {
+    /* Data structure */
+    struct aws_priority_queue queue;
+
+    /* Assumptions */
+    __CPROVER_assume(aws_priority_queue_is_bounded(&queue, MAX_INITIAL_ITEM_ALLOCATION, MAX_ITEM_SIZE));
+    bool backpointers_allocated = ensure_priority_queue_has_allocated_members(&queue);
+
+    /* Assume the function preconditions */
+    __CPROVER_assume(aws_priority_queue_is_valid(&queue));
+    void *item = can_fail_malloc(queue.container.item_size);
+
+    if (backpointers_allocated) {
+        /* Assume that the two backpointers 0, len-1 are valid,
+         * either by being NULL or by allocating their objects. This
+         * is important for the s_swap that happens in s_remove. */
+        size_t len = queue.backpointers.length;
+        if (0 < len) {
+            ((struct aws_priority_queue_node **)queue.backpointers.data)[0] =
+                can_fail_malloc(sizeof(struct aws_priority_queue_node));
+            if (0 != len - 1) {
+                ((struct aws_priority_queue_node **)queue.backpointers.data)[len - 1] =
+                    can_fail_malloc(sizeof(struct aws_priority_queue_node));
+            }
+        }
+    }
+
+    /* Perform operation under verification */
+    aws_priority_queue_pop(&queue, item);
+
+    /* Assert the postconditions */
+    assert(aws_priority_queue_is_valid(&queue));
```

Comment:
Are there any other assertions? For instance, check whether the size of the underlying `aws_array_list` decrease by 1. See `aws_array_list_pop_harness` for more possible assertions.

Reply:
I was thinking of keeping the first version of all harnesses in `priority_queue.c` about validity and have a separate pass later for proving functional correctness properties. Especially since a lot of functions such as `memcpy` and `sift` are stubbed out in these harnesses to make them run in reasonable time.